integrations/tensorflow/iree_tf_compiler/TF/SavedModelToIreeABI.cpp

```diff
@@ -0,0 +1,577 @@
+
+  // Add reflection attribute.
+  std::string refStr;
+  {
+    json::Object refDict;
+    refDict["a"] = json::Value(std::move(refArgs));
+    refDict["r"] = json::Value(std::move(refReturns));
+    json::Value refDictValue(std::move(refDict));
+    llvm::raw_string_ostream refOut(refStr);
+    refOut << refDictValue;
+    refOut.flush();
+  }
+
+  SmallVector<NamedAttribute> funcReflectAttrs(
+      {builder.getNamedAttr("abi", builder.getStringAttr("ireevm")),
+       builder.getNamedAttr("abiv", builder.getI32IntegerAttr(1)),
+       builder.getNamedAttr("d", builder.getStringAttr(refStr))});
+  wrapperFunc->setAttr("iree.reflection",
+                       builder.getDictionaryAttr(funcReflectAttrs));
+
+  return success();
+}
+
+}  // namespace
+
+class SavedModelToIreeABIPass
+    : public PassWrapper<SavedModelToIreeABIPass, OperationPass<ModuleOp>> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<iree_compiler::IREE::Flow::FlowDialect,
+                    iree_compiler::IREEDialect,
+                    mlir::tf_saved_model::TensorFlowSavedModelDialect>();
+  }
+
+  void runOnOperation() override {
+    if (failed(run())) {
+      signalPassFailure();
+    }
+  }
+
+  LogicalResult run() {
+    mlir::Builder builder(getOperation());
+    const Identifier savedModelIndexPathIdent =
+        builder.getIdentifier("tf_saved_model.index_path");
+    (void)savedModelIndexPathIdent;
+
+    // Handle saved model exported functions.
+    for (auto func : getOperation().getOps<FuncOp>()) {
+      // Transfer exported names to IREE.
+      auto exportedNames = mlir::tf_saved_model::GetExportedNames(func);
+      if (exportedNames.empty()) continue;
+      if (exportedNames.size() > 1) {
+        return func.emitError() << "Multiple exported names not supported yet";
+      }
+
+      StringRef exportedName = exportedNames.front();
+      StringRef internalName = func.getName();
+      if (internalName == exportedName) {
+        // Normally, the actual IR function name is some mangled form only
+        // relevant to some long departed TensorFlow devs. But there is nothing
+        // saying it has to be, so if there is a collision, be nice and move
+        // it out of the way.
+        std::string rename = internalName.str();
+        rename.append("__ireesm");
+        SymbolTable::setSymbolName(func, rename);
+      }
+      SymbolTable::setSymbolVisibility(func, SymbolTable::Visibility::Private);
+      if (failed(materializeABIWrapper(getOperation(), func, exportedName))) {
+        return failure();
+      }
+
+      // Remove its designation as a saved model export.
+      func->removeAttr("tf_saved_model.exported_names");
+    }
+
+    // We should have now removed anything requiring saved model semantics.
+    getOperation()->removeAttr("tf_saved_model.semantics");
+    return success();
+  }
+};
+
+std::unique_ptr<OperationPass<ModuleOp>> createSavedModelToIreeABIPass() {
```

Comment:
Would be nice to have consistent casing here - `IREEABI` is my preference (as it's `MLIR` not `Mlir`) - also happy to have a different more specific name if we want to name the calling convention. I use 'cconv' for short in some places but it could all use a consistency update at some point. At least for now not having the mixed casing would be nice :)